src/core/eme/create_session.ts

```diff
@@ -0,0 +1,179 @@
+import assert from "../../utils/assert";
+import castToObservable from "../../utils/castToObservable";
+import log from "../../utils/log";
+import { IMediaKeysInfos } from "./types";
+import isSessionUsable from "./utils/is_session_usable";
+
+export interface INewSessionCreatedEvent {
+  type : "created-session";
+  value : {
+    mediaKeySession : MediaKeySession|IMediaKeySession;
+    sessionType : MediaKeySessionType;
+  };
+}
+
+export interface IPersistentSessionRecoveryEvent {
+  type : "loaded-persistent-session";
+  value : {
+    mediaKeySession : MediaKeySession|IMediaKeySession;
+    sessionType : MediaKeySessionType;
+  };
+}
+
+export type ICreateSessionEvent =
+  INewSessionCreatedEvent |
+  IPersistentSessionRecoveryEvent;
+
+/**
+ * If session creating fails, retry once session creation/loading.
+ * Emit true, if it has succeeded to load, false if there is no data for the
+ * given sessionId.
+ * @param {string} sessionId
+ * @param {MediaKeySession} session
+ * @returns {Observable}
+ */
+function loadPersistentSession(
+  sessionId: string,
+  session: MediaKeySession|IMediaKeySession
+) : Observable<boolean> {
+  return Observable.defer(() => {
+    log.debug("eme: load persisted session", sessionId);
+    return castToObservable(session.load(sessionId));
+  });
+}
+
+/**
+ * Create a new Session on the given MediaKeys, corresponding to the given
+ * initializationData.
+ * If session creating fails, remove the oldest MediaKeySession loaded and
+ * retry.
+ *
+ * /!\ This only creates new sessions.
+ * It will fail if sessionsStore already has a MediaKeySession with
+ * the given initializationData.
+ * @param {Uint8Array} initData
+ * @param {string} initDataType
+ * @param {Object} mediaKeysInfos
+ * @returns {Observable}
+ */
+export default function createSession(
+  initData: Uint8Array,
+  initDataType: string,
+  mediaKeysInfos: IMediaKeysInfos
+) : Observable<ICreateSessionEvent> {
+  return Observable.defer(() => {
+    const {
+      keySystemOptions,
+      mediaKeySystemAccess,
+      sessionsStore,
+      sessionStorage,
+    } = mediaKeysInfos;
+
+    const mksConfig = mediaKeySystemAccess.getConfiguration();
+    const sessionTypes = mksConfig.sessionTypes;
+    const hasPersistence = (
+      sessionTypes && arrayIncludes(sessionTypes, "persistent-license")
+    );
+
+    const sessionType : MediaKeySessionType =
+      hasPersistence &&
+      sessionStorage &&
+      keySystemOptions.persistentLicense ?
+      "persistent-license" : "temporary";
+
+    log.debug(`eme: create a new ${sessionType} session`);
+
+    const session = sessionsStore.createSession(initData, initDataType, sessionType);
+
+    // Re-check for Dumb typescript
+    if (!hasPersistence || !sessionStorage || !keySystemOptions.persistentLicense) {
+      if (__DEV__) {
+        assert(sessionType === "temporary");
+      }
+      return Observable.of({
+        type: "created-session" as "created-session",
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Remove this unnecessary cast. ([typescript:S4325](https://sonarcloud.io/organizations/rx-player/rules#rule_key=typescript%3AS4325))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=rx-player&pullRequest=212&issues=AWNFqf_5XOBV_YrhBEpN&open=AWNFqf_5XOBV_YrhBEpN)

[ISSUE_KEY]: AWNFqf_5XOBV_YrhBEpN